{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "089c3dda961b404c96df0f91052cfd17",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>ipyTrenaViz</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "ipyTrenaViz()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import sys, time\n",
    "sys.path.append(\"/home/jovyan/mef2c\")\n",
    "from hub.Trena import *\n",
    "trena = Trena('hg38')\n",
    "trena.setWidgetHeight(600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'chrom': 'chr5', 'start': 88391000, 'end': 89322000}\n"
     ]
    }
   ],
   "source": [
    "trena.setWidgetHeight(800)\n",
    "print(trena.getBounds())\n",
    "trena.showGenomicRegion(trena.getBounds(asString=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Wishlist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load all of our data\n",
    "- display footprints for whole region, hint+wellington, 16+20\n",
    "- display enhancers, DHS\n",
    "- load and display eqtl variants\n",
    "- display vcf track\n",
    "- load pre-calculated models (we now have your 3 wg tcx, cer, ros models, +/-5bk, tfclass mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "tbl_fp = trena.getFootprintsInRegion(True)\n",
    "tbl_enhancers = trena.getEnhancersInRegion(True)\n",
    "tbl_var = trena.getWholeGenomeVariantsInRegion(1, 2, display=True, color=\"darkgreen\")\n",
    "tbl_var2 = trena.getWholeGenomeVariantsInRegion(altToRefRatio=2, minAltCount=5, display=True, color=\"magenta\")\n",
    "tbl_snp = trena.getVariants(minScore=0, display=True, color=\"red\") # , trackHeight=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## provide these functions\n",
    "- identifyCandidateVariants: parameterized by eqt|vcf, shoulder, background track"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## how to use getWholeGenomeVariantsInRegion\n",
    "\n",
    "This function takes \n",
    "  - an implicit argument, the genomic region currently displayed in igv (above): *roi* - the region of interest\n",
    "  - *altToRefRato*: a minimum proportion of AD/control samples with het or homozygous variants at each location\n",
    "  - *minAltCount*: the minimum number of AD (affected) samples at each location\n",
    "  - *display*: render the variants as a track in igv\n",
    "  - *color*: of the igv track\n",
    "  - *trackHeight*: 50 pixels by default"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['mef2c.cory.wgs.tcx.tfClass',\n",
       " 'mef2c.cory.wgs.cer.tfClass',\n",
       " 'mef2c.cory.wgs.ros.tfClass']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trena.getModelNames()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>beta.lasso</th>\n",
       "      <th>beta.ridge</th>\n",
       "      <th>rf.score</th>\n",
       "      <th>beta.sqrtlasso</th>\n",
       "      <th>lasso.p.value</th>\n",
       "      <th>pearson.coeff</th>\n",
       "      <th>spearman.coeff</th>\n",
       "      <th>concordance</th>\n",
       "      <th>pcaMax</th>\n",
       "      <th>target.gene</th>\n",
       "      <th>Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STAT4</td>\n",
       "      <td>0.1244</td>\n",
       "      <td>0.0516</td>\n",
       "      <td>21.6777</td>\n",
       "      <td>0.1606</td>\n",
       "      <td>2.431600e-15</td>\n",
       "      <td>0.4883</td>\n",
       "      <td>0.4724</td>\n",
       "      <td>0.5893</td>\n",
       "      <td>5.0723</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MEF2D</td>\n",
       "      <td>0.1252</td>\n",
       "      <td>0.0329</td>\n",
       "      <td>12.3736</td>\n",
       "      <td>0.1461</td>\n",
       "      <td>3.509900e-14</td>\n",
       "      <td>0.4799</td>\n",
       "      <td>0.4327</td>\n",
       "      <td>0.4969</td>\n",
       "      <td>3.6005</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SP8</td>\n",
       "      <td>0.0388</td>\n",
       "      <td>0.0184</td>\n",
       "      <td>8.8290</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.940200e-17</td>\n",
       "      <td>0.5240</td>\n",
       "      <td>0.4446</td>\n",
       "      <td>0.5829</td>\n",
       "      <td>3.3246</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SP9</td>\n",
       "      <td>0.0421</td>\n",
       "      <td>0.0245</td>\n",
       "      <td>9.0310</td>\n",
       "      <td>0.0323</td>\n",
       "      <td>2.782200e-15</td>\n",
       "      <td>0.5132</td>\n",
       "      <td>0.4283</td>\n",
       "      <td>0.5104</td>\n",
       "      <td>3.1168</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CUX2</td>\n",
       "      <td>0.1123</td>\n",
       "      <td>0.0543</td>\n",
       "      <td>4.7388</td>\n",
       "      <td>0.1277</td>\n",
       "      <td>2.954000e-12</td>\n",
       "      <td>0.4647</td>\n",
       "      <td>0.4093</td>\n",
       "      <td>0.5366</td>\n",
       "      <td>2.9343</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PAX7</td>\n",
       "      <td>0.1071</td>\n",
       "      <td>0.0307</td>\n",
       "      <td>4.5908</td>\n",
       "      <td>0.1280</td>\n",
       "      <td>2.308000e-14</td>\n",
       "      <td>0.5055</td>\n",
       "      <td>0.4323</td>\n",
       "      <td>0.5222</td>\n",
       "      <td>2.9173</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ZIC4</td>\n",
       "      <td>-0.0756</td>\n",
       "      <td>-0.0277</td>\n",
       "      <td>3.6807</td>\n",
       "      <td>-0.0471</td>\n",
       "      <td>1.380800e-14</td>\n",
       "      <td>-0.4917</td>\n",
       "      <td>-0.4390</td>\n",
       "      <td>0.4848</td>\n",
       "      <td>2.7780</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TBPL1</td>\n",
       "      <td>0.0185</td>\n",
       "      <td>0.0393</td>\n",
       "      <td>8.6657</td>\n",
       "      <td>0.0470</td>\n",
       "      <td>8.121200e-09</td>\n",
       "      <td>0.4507</td>\n",
       "      <td>0.3602</td>\n",
       "      <td>0.3976</td>\n",
       "      <td>2.5315</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BPTF</td>\n",
       "      <td>-0.0561</td>\n",
       "      <td>-0.0425</td>\n",
       "      <td>2.8119</td>\n",
       "      <td>-0.0379</td>\n",
       "      <td>8.527300e-10</td>\n",
       "      <td>-0.4364</td>\n",
       "      <td>-0.3602</td>\n",
       "      <td>0.3776</td>\n",
       "      <td>2.2793</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TEF</td>\n",
       "      <td>0.0192</td>\n",
       "      <td>0.0378</td>\n",
       "      <td>6.4553</td>\n",
       "      <td>0.0503</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.4039</td>\n",
       "      <td>0.3962</td>\n",
       "      <td>0.4311</td>\n",
       "      <td>2.0730</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>OLIG1</td>\n",
       "      <td>-0.0428</td>\n",
       "      <td>-0.0233</td>\n",
       "      <td>2.5015</td>\n",
       "      <td>-0.0377</td>\n",
       "      <td>5.326000e-08</td>\n",
       "      <td>-0.4253</td>\n",
       "      <td>-0.4007</td>\n",
       "      <td>0.3763</td>\n",
       "      <td>1.8898</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>DLX1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0194</td>\n",
       "      <td>5.5767</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>5.373000e-01</td>\n",
       "      <td>0.4846</td>\n",
       "      <td>0.4077</td>\n",
       "      <td>0.5248</td>\n",
       "      <td>1.7034</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NR3C2</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0283</td>\n",
       "      <td>5.1883</td>\n",
       "      <td>0.0270</td>\n",
       "      <td>3.000000e-04</td>\n",
       "      <td>0.2943</td>\n",
       "      <td>0.3565</td>\n",
       "      <td>0.4291</td>\n",
       "      <td>1.6102</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DLX2</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0144</td>\n",
       "      <td>4.0466</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>7.157000e-01</td>\n",
       "      <td>0.5171</td>\n",
       "      <td>0.4343</td>\n",
       "      <td>0.5040</td>\n",
       "      <td>1.6013</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>FOXN2</td>\n",
       "      <td>-0.0268</td>\n",
       "      <td>-0.0188</td>\n",
       "      <td>2.2615</td>\n",
       "      <td>-0.0507</td>\n",
       "      <td>6.132200e-06</td>\n",
       "      <td>-0.3412</td>\n",
       "      <td>-0.3450</td>\n",
       "      <td>0.3921</td>\n",
       "      <td>1.5544</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>DLX6</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0154</td>\n",
       "      <td>3.9344</td>\n",
       "      <td>0.0141</td>\n",
       "      <td>8.862000e-01</td>\n",
       "      <td>0.4954</td>\n",
       "      <td>0.4019</td>\n",
       "      <td>0.4955</td>\n",
       "      <td>1.5268</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ZIC1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.0193</td>\n",
       "      <td>3.4495</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>8.853000e-01</td>\n",
       "      <td>-0.4417</td>\n",
       "      <td>-0.4025</td>\n",
       "      <td>0.5029</td>\n",
       "      <td>1.5040</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ARID4B</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.0360</td>\n",
       "      <td>1.3159</td>\n",
       "      <td>-0.0036</td>\n",
       "      <td>5.330000e-02</td>\n",
       "      <td>-0.3521</td>\n",
       "      <td>-0.3247</td>\n",
       "      <td>0.4621</td>\n",
       "      <td>1.4750</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>BARHL2</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.0330</td>\n",
       "      <td>1.4788</td>\n",
       "      <td>-0.0385</td>\n",
       "      <td>3.000000e-04</td>\n",
       "      <td>-0.3115</td>\n",
       "      <td>-0.2704</td>\n",
       "      <td>0.4124</td>\n",
       "      <td>1.4683</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>FOXP1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0057</td>\n",
       "      <td>6.3552</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>8.734000e-01</td>\n",
       "      <td>0.3121</td>\n",
       "      <td>0.2009</td>\n",
       "      <td>0.4687</td>\n",
       "      <td>1.3348</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  beta.lasso  beta.ridge  rf.score  beta.sqrtlasso  lasso.p.value  \\\n",
       "1    STAT4      0.1244      0.0516   21.6777          0.1606   2.431600e-15   \n",
       "2    MEF2D      0.1252      0.0329   12.3736          0.1461   3.509900e-14   \n",
       "3      SP8      0.0388      0.0184    8.8290          0.0000   1.940200e-17   \n",
       "4      SP9      0.0421      0.0245    9.0310          0.0323   2.782200e-15   \n",
       "5     CUX2      0.1123      0.0543    4.7388          0.1277   2.954000e-12   \n",
       "6     PAX7      0.1071      0.0307    4.5908          0.1280   2.308000e-14   \n",
       "7     ZIC4     -0.0756     -0.0277    3.6807         -0.0471   1.380800e-14   \n",
       "8    TBPL1      0.0185      0.0393    8.6657          0.0470   8.121200e-09   \n",
       "9     BPTF     -0.0561     -0.0425    2.8119         -0.0379   8.527300e-10   \n",
       "10     TEF      0.0192      0.0378    6.4553          0.0503   0.000000e+00   \n",
       "11   OLIG1     -0.0428     -0.0233    2.5015         -0.0377   5.326000e-08   \n",
       "12    DLX1      0.0000      0.0194    5.5767          0.0000   5.373000e-01   \n",
       "13   NR3C2      0.0000      0.0283    5.1883          0.0270   3.000000e-04   \n",
       "14    DLX2      0.0000      0.0144    4.0466          0.0000   7.157000e-01   \n",
       "15   FOXN2     -0.0268     -0.0188    2.2615         -0.0507   6.132200e-06   \n",
       "16    DLX6      0.0000      0.0154    3.9344          0.0141   8.862000e-01   \n",
       "17    ZIC1      0.0000     -0.0193    3.4495          0.0000   8.853000e-01   \n",
       "18  ARID4B      0.0000     -0.0360    1.3159         -0.0036   5.330000e-02   \n",
       "19  BARHL2      0.0000     -0.0330    1.4788         -0.0385   3.000000e-04   \n",
       "20   FOXP1      0.0000      0.0057    6.3552          0.0000   8.734000e-01   \n",
       "\n",
       "    pearson.coeff  spearman.coeff  concordance  pcaMax target.gene  Rank  \n",
       "1          0.4883          0.4724       0.5893  5.0723       MEF2C     1  \n",
       "2          0.4799          0.4327       0.4969  3.6005       MEF2C     2  \n",
       "3          0.5240          0.4446       0.5829  3.3246       MEF2C     3  \n",
       "4          0.5132          0.4283       0.5104  3.1168       MEF2C     4  \n",
       "5          0.4647          0.4093       0.5366  2.9343       MEF2C     5  \n",
       "6          0.5055          0.4323       0.5222  2.9173       MEF2C     6  \n",
       "7         -0.4917         -0.4390       0.4848  2.7780       MEF2C     7  \n",
       "8          0.4507          0.3602       0.3976  2.5315       MEF2C     8  \n",
       "9         -0.4364         -0.3602       0.3776  2.2793       MEF2C     9  \n",
       "10         0.4039          0.3962       0.4311  2.0730       MEF2C    10  \n",
       "11        -0.4253         -0.4007       0.3763  1.8898       MEF2C    11  \n",
       "12         0.4846          0.4077       0.5248  1.7034       MEF2C    12  \n",
       "13         0.2943          0.3565       0.4291  1.6102       MEF2C    13  \n",
       "14         0.5171          0.4343       0.5040  1.6013       MEF2C    14  \n",
       "15        -0.3412         -0.3450       0.3921  1.5544       MEF2C    15  \n",
       "16         0.4954          0.4019       0.4955  1.5268       MEF2C    16  \n",
       "17        -0.4417         -0.4025       0.5029  1.5040       MEF2C    17  \n",
       "18        -0.3521         -0.3247       0.4621  1.4750       MEF2C    18  \n",
       "19        -0.3115         -0.2704       0.4124  1.4683       MEF2C    19  \n",
       "20         0.3121          0.2009       0.4687  1.3348       MEF2C    20  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tbl_model = trena.getModel('mef2c.cory.wgs.cer.tfClass')\n",
    "tfs = tbl_model.ix[:,\"gene\"].tolist()\n",
    "tbl_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## \"model threatening\" variants are calculated according to four independently varying parameters\n",
    "\n",
    "1. **chromosomal region**: inferred from the region currently displayed in igv\n",
    "2. **source of binding sites**: \"footprints\" *or* \"enhancer.motifs *or* \"dhs.motifs\" *or* \"allDNA.motifs\"\n",
    "3. **source of variants**: \"eqtl.snps\" *or* \"wg.AD.variants\"\n",
    "4. **motif-TF-mapping table**: \"MotifDb\" *or* \"TFClass\"\n",
    "5. **shoulder**: 0 or more bases \n",
    "6. **tfs**:  transcription factors to evaluate (gene symbols, usually drawn from the \"gene\" column of a trena model's table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### look for variants in binding sites of the first 5 tfs in the cerebral cortex model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>snpScore</th>\n",
       "      <th>strand</th>\n",
       "      <th>motifStart</th>\n",
       "      <th>motifEnd</th>\n",
       "      <th>motifName</th>\n",
       "      <th>motifScore</th>\n",
       "      <th>geneSymbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>88884587</td>\n",
       "      <td>88884587</td>\n",
       "      <td>rs3814428</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>+</td>\n",
       "      <td>88884580</td>\n",
       "      <td>88884591</td>\n",
       "      <td>Hsapiens-jaspar2016-SP8-MA0747.1</td>\n",
       "      <td>13.0111</td>\n",
       "      <td>SP8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5</td>\n",
       "      <td>88884587</td>\n",
       "      <td>88884587</td>\n",
       "      <td>rs3814428</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>-</td>\n",
       "      <td>88884581</td>\n",
       "      <td>88884589</td>\n",
       "      <td>Mmusculus-HOCOMOCOv10-ZIC1_MOUSE.H10MO.B</td>\n",
       "      <td>11.2778</td>\n",
       "      <td>ZIC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom     start       end         id  snpScore strand  motifStart  motifEnd  \\\n",
       "1  chr5  88884587  88884587  rs3814428    0.0218      +    88884580  88884591   \n",
       "2  chr5  88884587  88884587  rs3814428    0.0218      -    88884581  88884589   \n",
       "\n",
       "                                  motifName  motifScore geneSymbol  \n",
       "1          Hsapiens-jaspar2016-SP8-MA0747.1     13.0111        SP8  \n",
       "2  Mmusculus-HOCOMOCOv10-ZIC1_MOUSE.H10MO.B     11.2778       ZIC1  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trena.findVariantsInModel(\"footprints\", \"eqtl.snps\", tfs, \"TFClass\", 10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
